src/primitives/byte.cpp

```diff
@@ -0,0 +1,158 @@
+#include <cassert>
+#include <cstddef>
+
+#include "cubic-protocol/primitives/byte.hpp"
+
+namespace cubic::protocol::primitives::byte {
+
+auto serialize(std::vector<uint8_t> &out, int8_t value) -> void
+{
+    size_t current_size = out.size();
+    out.resize(out.size() + 1);
+    serialize(out.data() + current_size, value);
+}
+
+auto serialize(uint8_t *out, int8_t value) -> void
+{
+    // Magic pointer conversion bs
+    *out = *(uint8_t *) &value;
+}
+
+auto parse(const uint8_t *data, uint32_t available_bytes, int8_t *value) -> uint32_t
+{
+    if (available_bytes < 1)
+        return 0;
+    auto *casted_value = (uint8_t *) value;
+    *casted_value = *data;
+    return 1;
+}
+
+auto parse(const uint8_t *data, uint32_t available_bytes, int8_t *value, int8_t min, int8_t max)
+    -> uint32_t
+{
+    uint32_t parsed = parse(data, available_bytes, value);
+
+    if (parsed == 0 || *value < min || *value > max)
+        return 0;
+    return parsed;
+}
```

Comment:
min and max are signed while parsed is signed, this can cause undefined behavior when using negative min or max value

Reply:
`min` and `max` are not compared to `parsed` so this is not a problem.

Reply:
Yeah im dumb